**Testing Our Neural Network**

As for the code of our neural network we follow the book "Make Your Own Neural Network" from Tariq Rashid, 2016.
license is GPLv2

the original code can be found https://github.com/makeyourownneuralnetwork/makeyourownneuralnetwork

Testing is done using the mnist dataset, i.e., hand-written digit recognition.

Import libraries

In [1]:
import numpy
import time

load the mnist training data CSV file into a list

The list contains 100 entries

In [2]:
training_data_file = open("mnist_dataset/mnist_train_100.csv", 'r')
training_data_list = training_data_file.readlines()
training_data_file.close()

In [3]:
len(training_data_list)

100

Each list entry contains 785 numbers

The first number is the label and the 784 = 28X28 are the pixel values (greyscale 0-255)

In [4]:
training_data_list[0]

'5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,18,18,18,126,136,175,26,166,255,247,127,0,0,0,0,0,0,0,0,0,0,0,0,30,36,94,154,170,253,253,253,253,253,225,172,253,242,195,64,0,0,0,0,0,0,0,0,0,0,0,49,238,253,253,253,253,253,253,253,253,251,93,82,82,56,39,0,0,0,0,0,0,0,0,0,0,0,0,18,219,253,253,253,253,253,198,182,247,241,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,80,156,107,253,253,205,11,0,43,154,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,14,1,154,253,90,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,139,253,190,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,11,190,253,70,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,35,241,225,160,108,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,81,240,253,253,119,25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,

In [5]:
firstNumber = training_data_list[0].split(",")

In [6]:
len(firstNumber)

785

In [7]:
firstNumber

['5',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '3',
 '18',
 '18',
 '18',
 '126',
 '136',
 '175',
 '26',
 '166',
 '255',
 '247',


Create Neural Network

In [8]:
cd ../packageFlo

/home/florian/Projects/Own_Library/Own_Library/packageFlo


In [9]:
from snn import ShallowNeuralNetwork

In [10]:
help(ShallowNeuralNetwork)

Help on class ShallowNeuralNetwork in module snn:

class ShallowNeuralNetwork(builtins.object)
 |  ShallowNeuralNetwork(inputnodes, hiddennodes, outputnodes, learningrate)
 |  
 |  A class used to represent a simple feedforward neural network with one hidden layer.
 |  
 |  Detailed Explenation:
 |  ---------------------
 |  This is a code for a simple feedforward network with one hidden layer. It is a modification of the network found in the book
 |  "Make your own neural network", by T. Rashid. It works as following:
 |  To set up the network one specifies the number of input, hidden and output neurons and a learnig rate lr>0. Every neuron in
 |  one layer is connect to all the neurons in the next layer by a weight. Upon initialization the weights are (pseudo) randomly
 |  chosen between -0.5 and 0.5 by the random function.
 |  A neuron receives the weighted sum of all the neurons in the previous layer and applies the sigmoid function as the activation
 |  function. In this network w

In [11]:
help(ShallowNeuralNetwork.train)

Help on function train in module snn:

train(self, inputs_list, target_list)
    Training the network.
    
    This function trains the network using labeled examples, i.e.,
    the network weights get updated.
    
    Parameters:
    -----------
    input_list: float list
        This list represents one input for our network where each
        number stands for a feature of the input.
    target_list: float list
        This list contains the desired output for the given input
        object.



In [12]:
help(ShallowNeuralNetwork.get_output)

Help on function get_output in module snn:

get_output(self, inputs_list)
    Get output from neural network.
    
    This function takes an example as an input and returns an output
    calculated by the current state of the network.
    
    Parameters:
    -----------
    input_list: float list
        Each number in the list stands for a feature of an input object.
    
    Returns:
    --------
    final_outputs: float list
        List containing the calculated output.



In [13]:
# number of input, hidden and output nodes
input_nodes = 784
hidden_nodes = 200
output_nodes = 10

# learning rate
learning_rate = 0.1

# create instance of neural network
n = ShallowNeuralNetwork(input_nodes,hidden_nodes,output_nodes, learning_rate)

Train the neural network

In [14]:
st = time.time()

# epochs is the number of times the training data set is used for training
epochs = 2

for e in range(epochs):
    # go through all records in the training data set
    for record in training_data_list:
        # split the record by the ',' commas
        all_values = record.split(',')
        # scale and shift the inputs
        inputs = [(float(x)/255.0 * 0.99)+0.01 for x in all_values[1:]]
        # create the target output values (all 0.01, except the desired label which is 0.99)
        targets =  [0.01] * output_nodes
        # all_values[0] is the target label for this record
        targets[int(all_values[0])] = 0.99
        n.train(inputs, targets)
        pass
    pass

et = time.time()
print("Runtime:", et-st, "seconds")

Runtime: 1576.0858566761017 seconds


load the mnist test data CSV file into a list

In [15]:
cd ../ExamplesAndTests

/home/florian/Projects/Own_Library/Own_Library/ExamplesAndTests


In [16]:
test_data_file = open("mnist_dataset/mnist_test_10.csv", 'r')
test_data_list = test_data_file.readlines()
test_data_file.close()

test the neural network

In [17]:
# scorecard for how well the network performs, initially empty
scorecard = []

# go through all the records in the test data set
for record in test_data_list:
    # split the record by the ',' commas
    all_values = record.split(',')
    # correct answer is first value
    correct_label = int(all_values[0])
    # scale and shift the inputs
    inputs = ([(float(x)/255.0 * 0.99)+0.01 for x in all_values[1:]])
    # query the network
    outputs = n.get_output(inputs)
    # the index of the highest value corresponds to the label
    label = numpy.argmax(outputs)
    # append correct or incorrect to list
    if (label == correct_label):
        # network's answer matches correct answer, add 1 to scorecard
        scorecard.append(1)
    else:
        # network's answer doesn't match correct answer, add 0 to scorecard
        scorecard.append(0)
        pass
    
    pass

In [18]:
# calculate the performance score, the fraction of correct answers
scorecard_array = numpy.asarray(scorecard)
print ("performance = ", scorecard_array.sum() / scorecard_array.size)

performance =  0.6


test on training examples

In [19]:
# scorecard for how well the network performs, initially empty
scorecard = []

# go through all the records in the test data set
for record in training_data_list:
    # split the record by the ',' commas
    all_values = record.split(',')
    # correct answer is first value
    correct_label = int(all_values[0])
    # scale and shift the inputs
    inputs = ([(float(x)/255.0 * 0.99)+0.01 for x in all_values[1:]])
    # query the network
    outputs = n.get_output(inputs)
    # the index of the highest value corresponds to the label
    label = numpy.argmax(outputs)
    # append correct or incorrect to list
    if (label == correct_label):
        # network's answer matches correct answer, add 1 to scorecard
        scorecard.append(1)
    else:
        # network's answer doesn't match correct answer, add 0 to scorecard
        scorecard.append(0)
        pass
    
    pass


# calculate the performance score, the fraction of correct answers
scorecard_array = numpy.asarray(scorecard)
print ("performance = ", scorecard_array.sum() / scorecard_array.size)

performance =  0.96


We see above that the runtime for training the network is quite high

Below we want to investigate the reason for this

In [20]:
a = numpy.array([1,2,3,4])

In [21]:
a*a

array([ 1,  4,  9, 16])

In [22]:
x = [4] * 100000000

In [23]:
y = numpy.array(x)

In [24]:
st = time.time()

2*y*y+y


et = time.time()
print("Runtime:", et-st, "seconds")

Runtime: 0.33177685737609863 seconds


In [25]:
st = time.time()

[2*t*t+t for t in x]

et = time.time()
print("Runtime:", et-st, "seconds")

Runtime: 4.047400712966919 seconds


NumPy Arrays are faster than Python List comprehensions because of the following reasons:  

-An array is a collection of homogeneous data-types that are stored in contiguous memory locations. On the other hand, a list in Python is a collection of heterogeneous data types stored in non-contiguous memory locations.

-The NumPy package breaks down a task into multiple fragments and then processes all the fragments parallelly.

-The NumPy package integrates C, C++, and Fortran codes in Python. These programming languages have very little execution time compared to Python.

https://www.geeksforgeeks.org/why-numpy-is-faster-in-python/#:~:text=NumPy%20Arrays%20are%20faster%20than,in%20non%2Dcontiguous%20memory%20locations.


In our ShallowNeuralNetwork code we use lists instead of arrays from the numpy library. This makes all computations really slow.

**Conclusion:** Since we don't use the python package numpy we had to use python list comprehension in our snn module. This results in very high runtime.

We trained the network on 100 examples and testet it on 10 new examples. Because of the slow runtime we were only able to do 2 epochs. Nevertheless, the network labeled 6 out of 10 test examples correctly (and with more training it would probably be even better) which proves that the code works since the expactation value for random guesses would be 0.1. This is also supported by the fact that the network labes 96% of the training data correctly.